torch/csrc/jit/passes/utils/memory_dag.cpp

```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
   return all_a_mlocs.intersects(all_b_mlocs);
 }
 
-void MemoryDAG::makePointerTo(Element* from, Element* to) {
-  from->pointsTo.set(to->index);
-  from->cachedMemoryLocations_.clear();
-
-  to->pointedFrom.set(from->index);
+void MemoryDAGBuilder::makePointerTo(Element* from, Element* to) {
+  makePointerToImpl(from, to);
 }
 
-void MemoryDAG::addToContainedElements(Element* elem, Element* container) {
+void MemoryDAGBuilder::addToContainedElements(
+    Element* elem,
+    Element* container) {
   TORCH_INTERNAL_ASSERT(
       elem != container, "Elements cannot contain themselves");
   container->containedElements.set(elem->index);
 }
 
 // Give `v` a fresh alias (i.e. it does not point to any value)
-Element* MemoryDAG::makeFreshValue(const Value* v) {
-  indexToElementMap_.emplace_back(
-      torch::make_unique<Element>(*this, v, indexToElementMap_.size()));
-  return indexToElementMap_.back().get();
+Element* MemoryDAGBuilder::makeFreshValue(const Value* v) {
+  return makeFreshValueImpl(v, indexToElementMap_);
 }
 
-const MemoryLocations& Element::getMemoryLocations() const {
-  if (!cachedMemoryLocations_.empty()) {
-    return cachedMemoryLocations_;
+const MemoryLocations& MemoryDAG::getMemoryLocations(const Element* e) const {
+  if (e->cachedMemoryLocations_) {
+    return *e->cachedMemoryLocations_;
   }
 
-  // Do a BFS in the `points-to` direction, collecting all memory locations
   MemoryLocations ret;
-  this->bfs(BfsDirection::POINTS_TO, ret);
-  cachedMemoryLocations_ = ret;
-  return cachedMemoryLocations_;
-}
-
-// Do a breadth-first search over the graph, starting at `this` and
-// traversing in the direction `dir`.`fn` will be run on each element.
-void Element::bfs(BfsDirection dir, MemoryLocations& res) const {
-  std::queue<unsigned> queue;
-  ska::flat_hash_set<int> seen;
-  queue.push(this->index);
-  while (!queue.empty()) {
-    const auto index = queue.front();
-    queue.pop();
-    seen.insert(index);
-    auto el = dag.fromIndex(index);
-    if (el->pointsTo.empty()) {
-      res.set(index);
+  if (e->pointsTo.empty()) {
```

Comment:
It's a little surprising that we're not using a `seen` set in our DFS implementation, it may not trigger now but that's an easy way for an exponential runtime. Could you add a `getMemoryLocationsImpl` with a seen set ?

Reply:
Not sure what a seen set would do? If we have traversed `e` before, its `cachedMemoryLocations_` will be populated and thus we will return immediately from the traversal.

Reply:
yea nevermind, cachedMemoryLocations_ makes sure we don't redo work.